storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentReader.java

```diff
@@ -93,19 +92,23 @@ public long getNextIndex() {
   }
 
   @Override
-  public void reset(long index) {
+  public void reset(final long index) {
     reset();
-    Position position = this.index.lookup(index - 1);
-    if (position != null) {
+
+    final Position position = this.index.lookup(index - 1);
+    if (position != null && position.index() >= segment.index()) {
```

Comment:
So segment.index() is the first index? What happens when we over the last index?

Reply:
`JournalIndex#lookup(long)` returns the position of the given index or a lesser index, so it's not possible to get one that is outside of the segment (barring any bugs in `SegmentReader` which somehow routed a reset call to the wrong `FileChannelJournalSegmentReader`).

Reply:
That said, we could be more defensive and check it, it doesn't cost much

Reply:
So it would be possible to check whether we are in the right segment?

Reply:
Yes, the `JournalSegment#getLastIndex` returns the segment writer's last index (not the journal writer's last index), so barring any bugs, we can tell what is our first and last index.

Like I said, this is already done by the caller, but it doesn't cost much to do here either. I lean towards leaving it out, but I have no strong arguments so I'll defer to your decision :+1:

Reply:
I think if we check the first then we should also check the last - bugs happen